File: go/heap.go
package main

import (
	"errors"
)

type MinHeap[T ordered] struct {
	heap []T
}

func NewMinHeap[T ordered]() *MinHeap[T] {
	return &MinHeap[T]{heap: make([]T, 0)}
}

func (mh *MinHeap[T]) Size() int {
	return len(mh.heap)
}

func (mh *MinHeap[T]) IsEmpty() bool {
	return len(mh.heap) == 0
}

func (mh *MinHeap[T]) getParentIndex(index int) int {
	return (index - 1) / 2
}

func (mh *MinHeap[T]) getLeftChildIndex(index int) int {
	return 2*index + 1
}

func (mh *MinHeap[T]) getRightChildIndex(index int) int {
	return 2*index + 2
}

func (mh *MinHeap[T]) hasParent(index int) bool {
	return mh.getParentIndex(index) >= 0
}

func (mh *MinHeap[T]) hasLeftChild(index int) bool {
	return mh.getLeftChildIndex(index) < len(mh.heap)
}

func (mh *MinHeap[T]) hasRightChild(index int) bool {
	return mh.getRightChildIndex(index) < len(mh.heap)
}

func (mh *MinHeap[T]) parent(index int) T {
	return mh.heap[mh.getParentIndex(index)]
}

func (mh *MinHeap[T]) leftChild(index int) T {
	return mh.heap[mh.getLeftChildIndex(index)]
}

func (mh *MinHeap[T]) rightChild(index int) T {
	return mh.heap[mh.getRightChildIndex(index)]
}

func (mh *MinHeap[T]) swap(indexOne, indexTwo int) {
	mh.heap[indexOne], mh.heap[indexTwo] = mh.heap[indexTwo], mh.heap[indexOne]
}

func (mh *MinHeap[T]) heapifyUp() {
	index := len(mh.heap) - 1
	for mh.hasParent(index) && mh.parent(index) > mh.heap[index] {
		mh.swap(mh.getParentIndex(index), index)
		index = mh.getParentIndex(index)
	}
}

func (mh *MinHeap[T]) heapifyDown() {
	index := 0
	for mh.hasLeftChild(index) {
		smallerChildIndex := mh.getLeftChildIndex(index)
		if mh.hasRightChild(index) && mh.rightChild(index) < mh.leftChild(index) {
			smallerChildIndex = mh.getRightChildIndex(index)
		}

		if mh.heap[index] < mh.heap[smallerChildIndex] {
			break
		} else {
			mh.swap(index, smallerChildIndex)
		}
		index = smallerChildIndex
	}
}

func (mh *MinHeap[T]) Peek() (T, error) {
	var zero T
	if mh.IsEmpty() {
		return zero, errors.New("heap is empty")
	}
	return mh.heap[0], nil
}

func (mh *MinHeap[T]) Poll() (T, error) {
	var zero T
	if mh.IsEmpty() {
		return zero, errors.New("heap is empty")
	}
	item := mh.heap[0]
	lastIndex := len(mh.heap) - 1
	lastItem := mh.heap[lastIndex]
	mh.heap = mh.heap[:lastIndex]
	if len(mh.heap) > 0 {
		mh.heap[0] = lastItem
		mh.heapifyDown()
	}
	return item, nil
}

func (mh *MinHeap[T]) Add(item T) {
	mh.heap = append(mh.heap, item)
	mh.heapifyUp()
}

func (mh *MinHeap[T]) ToSlice() []T {
	result := make([]T, len(mh.heap))
	copy(result, mh.heap)
	return result
}

func (mh *MinHeap[T]) Clear() {
	mh.heap = make([]T, 0)
}

type MaxHeap[T ordered] struct {
	heap []T
}

func NewMaxHeap[T ordered]() *MaxHeap[T] {
	return &MaxHeap[T]{heap: make([]T, 0)}
}

func (mh *MaxHeap[T]) Size() int {
	return len(mh.heap)
}

func (mh *MaxHeap[T]) IsEmpty() bool {
	return len(mh.heap) == 0
}

func (mh *MaxHeap[T]) getParentIndex(index int) int {
	return (index - 1) / 2
}

func (mh *MaxHeap[T]) getLeftChildIndex(index int) int {
	return 2*index + 1
}

func (mh *MaxHeap[T]) getRightChildIndex(index int) int {
	return 2*index + 2
}

func (mh *MaxHeap[T]) hasParent(index int) bool {
	return mh.getParentIndex(index) >= 0
}

func (mh *MaxHeap[T]) hasLeftChild(index int) bool {
	return mh.getLeftChildIndex(index) < len(mh.heap)
}

func (mh *MaxHeap[T]) hasRightChild(index int) bool {
	return mh.getRightChildIndex(index) < len(mh.heap)
}

func (mh *MaxHeap[T]) parent(index int) T {
	return mh.heap[mh.getParentIndex(index)]
}

func (mh *MaxHeap[T]) leftChild(index int) T {
	return mh.heap[mh.getLeftChildIndex(index)]
}

func (mh *MaxHeap[T]) rightChild(index int) T {
	return mh.heap[mh.getRightChildIndex(index)]
}

func (mh *MaxHeap[T]) swap(indexOne, indexTwo int) {
	mh.heap[indexOne], mh.heap[indexTwo] = mh.heap[indexTwo], mh.heap[indexOne]
}

func (mh *MaxHeap[T]) heapifyUp() {
	index := len(mh.heap) - 1
	for mh.hasParent(index) && mh.parent(index) < mh.heap[index] {
		mh.swap(mh.getParentIndex(index), index)
		index = mh.getParentIndex(index)
	}
}

func (mh *MaxHeap[T]) heapifyDown() {
	index := 0
	for mh.hasLeftChild(index) {
		largerChildIndex := mh.getLeftChildIndex(index)
		if mh.hasRightChild(index) && mh.rightChild(index) > mh.leftChild(index) {
			largerChildIndex = mh.getRightChildIndex(index)
		}

		if mh.heap[index] > mh.heap[largerChildIndex] {
			break
		} else {
			mh.swap(index, largerChildIndex)
		}
		index = largerChildIndex
	}
}

func (mh *MaxHeap[T]) Peek() (T, error) {
	var zero T
	if mh.IsEmpty() {
		return zero, errors.New("heap is empty")
	}
	return mh.heap[0], nil
}

func (mh *MaxHeap[T]) Poll() (T, error) {
	var zero T
	if mh.IsEmpty() {
		return zero, errors.New("heap is empty")
	}
	item := mh.heap[0]
	lastIndex := len(mh.heap) - 1
	lastItem := mh.heap[lastIndex]
	mh.heap = mh.heap[:lastIndex]
	if len(mh.heap) > 0 {
		mh.heap[0] = lastItem
		mh.heapifyDown()
	}
	return item, nil
}

func (mh *MaxHeap[T]) Add(item T) {
	mh.heap = append(mh.heap, item)
	mh.heapifyUp()
}

func (mh *MaxHeap[T]) ToSlice() []T {
	result := make([]T, len(mh.heap))
	copy(result, mh.heap)
	return result
}

func (mh *MaxHeap[T]) Clear() {
	mh.heap = make([]T, 0)
}

File: dart/heap.dart
class MinHeap<T extends Comparable> {
  final List<T> _heap = [];

  int get size => _heap.length;
  bool get isEmpty => _heap.isEmpty;

  int _getParentIndex(int index) => (index - 1) ~/ 2;
  int _getLeftChildIndex(int index) => 2 * index + 1;
  int _getRightChildIndex(int index) => 2 * index + 2;
  bool _hasParent(int index) => _getParentIndex(index) >= 0;
  bool _hasLeftChild(int index) => _getLeftChildIndex(index) < _heap.length;
  bool _hasRightChild(int index) => _getRightChildIndex(index) < _heap.length;
  T _parent(int index) => _heap[_getParentIndex(index)];
  T _leftChild(int index) => _heap[_getLeftChildIndex(index)];
  T _rightChild(int index) => _heap[_getRightChildIndex(index)];

  void _swap(int indexOne, int indexTwo) {
    T temp = _heap[indexOne];
    _heap[indexOne] = _heap[indexTwo];
    _heap[indexTwo] = temp;
  }

  void _heapifyUp() {
    int index = _heap.length - 1;
    while (_hasParent(index) && _parent(index).compareTo(_heap[index]) > 0) {
      _swap(_getParentIndex(index), index);
      index = _getParentIndex(index);
    }
  }

  void _heapifyDown() {
    int index = 0;
    while (_hasLeftChild(index)) {
      int smallerChildIndex = _getLeftChildIndex(index);
      if (_hasRightChild(index) && _rightChild(index).compareTo(_leftChild(index)) < 0) {
        smallerChildIndex = _getRightChildIndex(index);
      }

      if (_heap[index].compareTo(_heap[smallerChildIndex]) < 0) {
        break;
      } else {
        _swap(index, smallerChildIndex);
      }
      index = smallerChildIndex;
    }
  }

  T peek() {
    if (isEmpty) throw StateError('heap is empty');
    return _heap[0];
  }

  T poll() {
    if (isEmpty) throw StateError('heap is empty');
    T item = _heap[0];
    int lastIndex = _heap.length - 1;
    T lastItem = _heap[lastIndex];
    _heap.removeLast();
    if (_heap.isNotEmpty) {
      _heap[0] = lastItem;
      _heapifyDown();
    }
    return item;
  }

  void add(T item) {
    _heap.add(item);
    _heapifyUp();
  }

  void insert(T item) {
    add(item);
  }

  List<T> toSlice() {
    return List.from(_heap);
  }

  void clear() {
    _heap.clear();
  }
}

class MaxHeap<T extends Comparable> {
  final List<T> _heap = [];

  int get size => _heap.length;
  bool get isEmpty => _heap.isEmpty;

  int _getParentIndex(int index) => (index - 1) ~/ 2;
  int _getLeftChildIndex(int index) => 2 * index + 1;
  int _getRightChildIndex(int index) => 2 * index + 2;
  bool _hasParent(int index) => _getParentIndex(index) >= 0;
  bool _hasLeftChild(int index) => _getLeftChildIndex(index) < _heap.length;
  bool _hasRightChild(int index) => _getRightChildIndex(index) < _heap.length;
  T _parent(int index) => _heap[_getParentIndex(index)];
  T _leftChild(int index) => _heap[_getLeftChildIndex(index)];
  T _rightChild(int index) => _heap[_getRightChildIndex(index)];

  void _swap(int indexOne, int indexTwo) {
    T temp = _heap[indexOne];
    _heap[indexOne] = _heap[indexTwo];
    _heap[indexTwo] = temp;
  }

  void _heapifyUp() {
    int index = _heap.length - 1;
    while (_hasParent(index) && _parent(index).compareTo(_heap[index]) < 0) {
      _swap(_getParentIndex(index), index);
      index = _getParentIndex(index);
    }
  }

  void _heapifyDown() {
    int index = 0;
    while (_hasLeftChild(index)) {
      int largerChildIndex = _getLeftChildIndex(index);
      if (_hasRightChild(index) && _rightChild(index).compareTo(_leftChild(index)) > 0) {
        largerChildIndex = _getRightChildIndex(index);
      }

      if (_heap[index].compareTo(_heap[largerChildIndex]) > 0) {
        break;
      } else {
        _swap(index, largerChildIndex);
      }
      index = largerChildIndex;
    }
  }

  T peek() {
    if (isEmpty) throw StateError('heap is empty');
    return _heap[0];
  }

  T poll() {
    if (isEmpty) throw StateError('heap is empty');
    T item = _heap[0];
    int lastIndex = _heap.length - 1;
    T lastItem = _heap[lastIndex];
    _heap.removeLast();
    if (_heap.isNotEmpty) {
      _heap[0] = lastItem;
      _heapifyDown();
    }
    return item;
  }

  void add(T item) {
    _heap.add(item);
    _heapifyUp();
  }

  void insert(T item) {
    add(item);
  }

  List<T> toSlice() {
    return List.from(_heap);
  }

  void clear() {
    _heap.clear();
  }
}
